README.md
# russian_license_plate

Input and output widgets for russian cars' license plates.

License plates types:
- **Standart** for regular cars. Examle: `В257УК 27`. 
- **Route** for taxis and buses. Examle: `TA001 125`.
- **Police** for polices casr. Examle: `M0001 716`.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/full_example.gif?raw=true)

## Getting Started 🛠

- Add the following dependency to pubspec.yaml of your project: 

```
dependencies:
  russian_license_plate: ^0.0.1
```

- Press `Pub get` (for Android Studio) or run `flutter packages upgrade`

- Import `plates.dart` to your code.

```
import 'package:russian_license_plate/plates.dart';
```

## Usage 

### Standart output 🚗

License plate output for regular cars.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/standart_output.png?raw=true)

***Class***: **`StandartLicensePlateOutput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`licensePlate`** (`StandartLicensePlate`) - output license plate
- **`showFlag`** (`bool`) - show russian flag under region number if true, default value is `true`

***Code example:***
```dart
StandartLicensePlateOutput(
    size: 70,
    licensePlate: StandartLicensePlate.from("В257УК27"),
)
```

### Standart input 🚘

License plate input for regular cars.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/standart_input.png?raw=true)

***Class***: **`StandartLicensePlateInput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`hintPlate`** (`StandartLicensePlate?`) - input hint license plate, default value is `B257YK27`
- **`showFlag`** (`bool`) - show russian flag under region number if true, default value is `true`
- **`onSubmitted`** (`Function(StandartLicensePlate?, LicensePlateInputError?)?`) - called when user end entered

***Code example:***
```dart
StandartLicensePlateInput(
    size: 70,
    onSubmitted: (value, error) {
        print("$value $error");
    },
)
```

### Route output 🚕

License plate output for taxis and buses.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/route_output.png?raw=true)

***Class***: **`RouteLicensePlateOutput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`licensePlate`** (`RouteLicensePlate`) - output license plate

***Code example:***
```dart
RouteLicensePlateOutput(
    size: 70,
    licensePlate: RouteLicensePlate.from("ТА010116"),
)
```

### Route input 🚖

License plate input for taxis and buses.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/route_input.png?raw=true) 

***Class***: **`RouteLicensePlateInput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`hintPlate`** (`RouteLicensePlate?`) - input hint license plate, default value is `TA777125`
- **`onSubmitted`** (`Function(RouteLicensePlate?, LicensePlateInputError?)?`) - called when user end entered

***Code example:***
```dart
RouteLicensePlateInput(
    size: 70,
    onSubmitted: (value, error) {
        print("$value $error");
    },
)
```

### Police output 🚓

License plate output for police cars.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/police_output.png?raw=true)

***Class***: **`PoliceLicensePlateOutput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`licensePlate`** (`PoliceLicensePlate`) - output license plate

***Code example:***
```dart
PoliceLicensePlateOutput(
    size: 70,
    licensePlate: PoliceLicensePlate.from("M6666716"),
)
```

### Police input 🚔

License plate input for police cars.

![Image](https://github.com/InnoFlutter/russian_license_plate/blob/main/img/police_input.png?raw=true)

***Class***: **`PoliceLicensePlateInput`**

***Parameters***:
- **`size`** (`double`) - size of the widget
- **`hintPlate`** (`PoliceLicensePlate?`) - input hint license plate, default value is `M000125`
- **`onSubmitted`** (`Function(PoliceLicensePlate?, LicensePlateInputError?)?`) - called when user end entered

***Code example:***
```dart
PoliceLicensePlateInput(
    size: 70,
    onSubmitted: (value, error) {
        print("$value $error");
    },
)
```

## License ✍️

We use [MIT licence](LICENSE).

## Support ❤️🤑

You can **[support me here](https://www.tinkoff.ru/cf/2HnHfETmeEv)**.

## Package 🧃

You can find the package [here](https://pub.dev/packages/russian_license_plate)


CHANGELOG.md
## 0.0.1

* LicensePlates: Initial release.

## 0.0.2

* Descriptions: Updated.

## 0.0.3

* Images: Updated.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:russian_license_plate/plates.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        // appBar: AppBar(
        //   title: const Text('Plugin example app'),
        // ),
        body: SingleChildScrollView(child: _body()),
      ),
    );
  }

  Widget _body() {
    return Container(
      alignment: Alignment.topCenter,
      child: Wrap(
        children: [
          _makeBlock(
            input: _standartLicensePlateInput(),
            outputMaker: _standartLicensePlateOutput,
            error: _standartLicensePlateInputError,
          ),
          _makeBlock(
            input: _routeLicensePlateInput(),
            outputMaker: _routeLicensePlateOutput,
            error: _routeLicensePlateInputError,
          ),
          _makeBlock(
            input: _policeLicensePlateInput(),
            outputMaker: _policeLicensePlateOutput,
            error: _policeLicensePlateInputError,
          ),
        ],
      ),
    );
  }

  /*******************--[Standart cars license plate usage]--*******************/
  /**
   * License plate for regulars cars.
   * Examples: "В257УК 27", "Р474АО 125"
   * **/

  /// Create a standart license plate "P474AO 125"
  StandartLicensePlate _standartLicensePlateValue =
      StandartLicensePlate.from("P474AO125");

  Widget _standartLicensePlateOutput(double size) {
    /// Standart license plate output
    ///
    /// - size: widget size
    /// - licensePlate: plate model
    return StandartLicensePlateOutput(
      size: size,
      licensePlate: _standartLicensePlateValue,
    );
  }

  /// Input error
  LicensePlateInputError? _standartLicensePlateInputError;

  Widget _standartLicensePlateInput() {
    /// Standart cars license plate input
    ///
    /// - size: widget size
    /// - onSubmitted: called when user entered license plate
    ///   - value: license plate model
    ///   - error: input error
    return StandartLicensePlateInput(
      size: 75,
      onSubmitted: (value, error) {
        if (value != null) {
          _standartLicensePlateValue = value;
        }
        _standartLicensePlateInputError = error;
        setState(() {});
      },
    );
  }

  /*******************--[Standart cars license plate usage]--*******************/

  /*******************--[Route cars license plate usage]--*******************/
  /**
   * License plate for taxis and buses.
   * Examples: "ТА 001 27", "АА 123 125"
   * **/

  /// Create a route license plate "TA 001 27"
  RouteLicensePlate _routeLicensePlateValue = RouteLicensePlate.from("TA00127");

  Widget _routeLicensePlateOutput(double size) {
    /// Route license plate output
    ///
    /// - size: widget size
    /// - licensePlate: plate model
    return RouteLicensePlateOutput(
      size: size,
      licensePlate: _routeLicensePlateValue,
    );
  }

  /// Input error
  LicensePlateInputError? _routeLicensePlateInputError;

  Widget _routeLicensePlateInput() {
    /// Route cars license plate input
    ///
    /// - size: widget size
    /// - onSubmitted: called when user entered license plate
    ///   - value: license plate model
    ///   - error: input error
    return RouteLicensePlateInput(
      size: 75,
      onSubmitted: (value, error) {
        if (value != null) {
          _routeLicensePlateValue = value;
        }
        _routeLicensePlateInputError = error;
        setState(() {});
      },
    );
  }

  /*******************--[Route cars license plate usage]--*******************/

  /*******************--[Police cars license plate usage]--*******************/
  /**
   * License plate for taxis and buses.
   * Examples: "М 6666 716", "А 0001 16"
   * **/

  /// Create a police license plate "M 6666 716"
  PoliceLicensePlate _policeLicensePlateValue =
      PoliceLicensePlate.from("M6666716");

  Widget _policeLicensePlateOutput(double size) {
    /// Police license plate output
    ///
    /// - size: widget size
    /// - licensePlate: plate model
    return PoliceLicensePlateOutput(
      size: size,
      licensePlate: _policeLicensePlateValue,
    );
  }

  /// Input error
  LicensePlateInputError? _policeLicensePlateInputError;

  Widget _policeLicensePlateInput() {
    /// Police cars license plate input
    ///
    /// - size: widget size
    /// - onSubmitted: called when user entered license plate
    ///   - value: license plate model
    ///   - error: input error
    return PoliceLicensePlateInput(
      size: 75,
      onSubmitted: (value, error) {
        if (value != null) {
          _policeLicensePlateValue = value;
        }
        _policeLicensePlateInputError = error;
        setState(() {});
      },
    );
  }

  /*******************--[Police cars license plate usage]--*******************/

  /// Make block with a input and different size outputs
  Widget _makeBlock({
    required Widget input,
    required LicensePlateInputError? error,
    required Widget Function(double) outputMaker,
  }) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Column(
        children: [
              input,
              _showInputErrorFor(error),
            ] +
            sizes
                .map((size) => outputMaker(size))
                .toList()
                .map((e) => Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: e,
                    ))
                .toList(),
      ),
    );
  }

  /// Sizes for outputs
  List<double> sizes = [75, 50, 25];

  /// Error widget
  Widget _showInputErrorFor(LicensePlateInputError? error) {
    if (error == null) {
      return const Text(
        "No errors",
        style: TextStyle(color: Colors.green),
      );
    }

    String errorIn;
    switch (error) {
      case LicensePlateInputError.region:
        errorIn = "region";
        break;
      case LicensePlateInputError.series:
        errorIn = "series";
        break;
      case LicensePlateInputError.number:
        errorIn = "number";
        break;
    }
    return Text(
      "Error in $errorIn",
      style: const TextStyle(color: Colors.red),
    );
  }
}
